src/ext/coms.c

```diff
@@ -0,0 +1,440 @@
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "coms.h"
+#include "env_config.h"
+#include "mpack.h"
+#include "vendor_stdatomic.h"
+
+typedef uint32_t group_id_t;
+
+#define GROUP_ID_PROCESSED (1UL << 31UL)
+
+#ifndef __clang__
+// disable checks since older GCC is throwing false errors
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
+ddtrace_coms_state_t ddtrace_coms_global_state = {{0}};
+#pragma GCC diagnostic pop
+#else  //__clang__
+ddtrace_coms_state_t ddtrace_coms_global_state = {.stacks = NULL};
+#endif
+
+static uint32_t store_data(group_id_t group_id, const char *src, size_t size) {
+    ddtrace_coms_stack_t *stack = atomic_load(&ddtrace_coms_global_state.current_stack);
+    if (stack == NULL) {
+        // no stack to save data to
+        return ENOMEM;
+    }
+
+    size_t size_to_alloc = size + sizeof(size_t) + sizeof(group_id_t);
+
+    atomic_fetch_add(&stack->refcount, 1);
+
+    size_t position = atomic_fetch_add(&stack->position, size_to_alloc);
+    if ((position + size_to_alloc) > stack->size) {
+        // allocation failed
+        atomic_fetch_sub(&stack->refcount, 1);
+        return ENOMEM;
+    }
+
+    memcpy(stack->data + position, &size, sizeof(size_t));
+    position += sizeof(size_t);
+
+    memcpy(stack->data + position, &group_id, sizeof(group_id_t));
+    position += sizeof(group_id_t);
+
+    memcpy(stack->data + position, src, size);
+
+    atomic_fetch_add(&stack->bytes_written, size_to_alloc);
+    atomic_fetch_sub(&stack->refcount, 1);
+    return 0;
+}
+
+ddtrace_coms_stack_t *new_stack() {
+    ddtrace_coms_stack_t *stack = calloc(1, sizeof(ddtrace_coms_stack_t));
```

Comment:
For all the `calloc()`'s and `malloc()`'s, is there a reason why we're not going through the ZMM? It might help reduce the number of memory allocations using `ecalloc()` and `emalloc()` (unless I missed something which is totally possible). :)

Reply:
We're intentionally avoiding ZMM :)  
This way we're not constrained by the arbitrarily set PHP memory limits. Also this data is shared between requests, I'm not even sure how ZMM tackles that :)

Reply:
Makes total sense! For future reference, if we want to persist memory in ZMM across requests we could use `pecalloc()` and `pemalloc()`, but we won't need to worry about that for now. :)